Feature/docker improvements

tomereli opened the pull request:
Implement docker improvements according to #72 

- [x] Add an option to use dynamic base image based on the host (-n / --native)
- [x] Update run.sh to call image build if needed
- [x] Add docker-builder.sh and docker-runner.sh helper scripts

Commit:
tools: docker runner - build image automatically

Build docker image prplmesh-runner when run.sh is called in case image
is not built.

See issue #72

Signed-off-by: tbeliyah <tomer.b.eliyahu@intel.com>

Commit:
tools: docker builder - build image automatically

Automatically call image-build.sh from build.sh with default parameters if the
prplmesh-build image does not exist.

See issue #72

Signed-off-by: tbeliyah <tomer.b.eliyahu@intel.com>

Commit:
tools: docker builder build native os option

In some cases we want to build using the same OS version as in the host.
For example, consider the case that the host is ubuntu-19.04, where some
of the packages which are installed are newer than the default image
from which the builder derives (ubuntu-18.04).
This means that we won't necessarilly be able to run the images built by
the builder from the host.

This is a major issue though since we can also run from docker using the
runner docker, but adding an option to build the base image using the
same OS version is nice, so this commit adds it.

See issue #72

Signed-off-by: tbeliyah <tomer.b.eliyahu@intel.com>

Commit:
tools: docker runner: option to use host os

Add an option to build an image based on the host OS version instead of
the hard coded Ubuntu-18.04.

This is needed in some cases when we want to be able to run prplMesh
built on the host, so the shared libraries (for example json-c) can
differ between ubuntu versions.

This commit adds an optional flag to the image-build.sh script to use
the host OS version instead of always using ubuntu-18.04.

Signed-off-by: tbeliyah <tomer.b.eliyahu@intel.com>

Commit:
tools: add docker helper scripts

Add docker-builder.sh and docker-runner.sh helper scripts.
This will make it more obvious on how to build and how to run prplMesh
since the build and run scripts are at the top level directory.

Signed-off-by: tbeliyah <tomer.b.eliyahu@intel.com>

Commit:
tools: docker runner add gdb

Signed-off-by: tbeliyah <tomer.b.eliyahu@intel.com>

Commit:
tools: docker runner: remove installdir from image

Having the install directory in the image may cause accidently
using the wrong directory as WORKDIR in case the image was
built on a different prplMesh clone.

Remove installdir from the image, move it to the entry point.
In addition, add environment variable INSTALL_DIR to point on the actual
install dir and use it in the helper scripts start-agent, start-
controller and start-controller-agent.

Signed-off-by: tbeliyah <tomer.b.eliyahu@intel.com>

arnout commented on tools/docker/builder/image-build.sh:
```diff
@@ -1,20 +1,64 @@
+    done
+
+    [ "$NATIVE" = "true" ] && {
+        . /etc/lsb-release
```
`lsb-release` is pretty deprecated AFAIK. At least, I don't have it. Use `/etc/os-release` instead.

arnout commented on tools/docker/builder/image-build.sh:
```diff
@@ -1,20 +1,64 @@
+        esac
+    done
+
+    [ "$NATIVE" = "true" ] && {
```
Since you're going to source a file below, I think it would be more appropriate to do this in a sub-shell, i.e. `(` instead of `{`

arnout commented on tools/docker/builder/image-build.sh:
```diff
@@ -1,20 +1,64 @@
+
+    [ "$NATIVE" = "true" ] && {
+        . /etc/lsb-release
+        IMAGE=$(echo $DISTRIB_ID | awk '{print tolower($0)}'):$DISTRIB_RELEASE
```
DISTRIB_ID and DISTRIB_RELEASE are not standard either (at least, they don't appear in my Fedora's os-release). Use `ID` and `VERSION_ID` from os-release instead.

arnout commented on docker-builder.sh:
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
+
+scriptdir="$(cd "${0%/*}"; pwd)"
+${scriptdir}/tools/docker/builder/build.sh "$@"
```
I would actually prefer a single `docker.sh` script that takes build/run as an argument.

That would probably require changing the names of the directories, to make it easier. I.e. `tools/docker/build/build.sh`.

tomereli replied:
Good point, done.

arnout reviewed (approved):
As usual, my comments are not critical so I approve already.

Commit:
tools: docker unify image-build.sh

Unify image-build.sh to support building both prplmesh-build and
prplmesh-runner docker images.

Signed-off-by: tbeliyah <tomer.b.eliyahu@intel.com>

Commit:
docker: image-build.sh base image option

Add option for selecting the builder and runner base image
Used for the Dockerfile FROM command, and set the default according to
the host (/etc/os-release).

Signed-off-by: tbeliyah <tomer.b.eliyahu@intel.com>
